Implement `arrow3d` method for `Axes3D`

AdwaithBatchu opened the pull request:
<!--
Thank you so much for your PR!  To help us review your contribution, please check
out the development guide https://matplotlib.org/devdocs/devel/index.html
-->

## PR summary
<!-- Please describe the pull request, using the questions below as guidance, and link to any relevant issues and PRs:

- Why is this change necessary?
- What problem does it solve?
- What is the reasoning for this implementation?

Additionally, please summarize the changes in the title, for example "Raise ValueError on
non-numeric input to set_xlim" and avoid non-descriptive titles such as "Addresses
issue #8576".

If possible, please provide a minimum self-contained example.  If you have used
generative AI as an aid in preparing this PR, see

https://matplotlib.org/devdocs/devel/contribute.html#restrictions-on-generative-ai-usage
-->

closes 22571

This PR implements a method to plot a 3D arrow.

The `ax.arrow3d` method plots a single arrow between specified end, start points. It returns an instance of the `Arrow3D` class,  which inherits from the `FancyArrowPatch` class. Additional kwargs are passed on to `FancyArrowPatch` to change properties of the arrow.
## PR checklist
<!-- Please mark any checkboxes that do not apply to this PR as [N/A].-->

- [x] "closes #0000" is in the body of the PR description to [link the related issue](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue)
- [x] new and changed code is [tested](https://matplotlib.org/devdocs/devel/testing.html)
- [x] *Plotting related* features are demonstrated in an [example](https://matplotlib.org/devdocs/devel/document.html#write-examples-and-tutorials)
- [x] *New Features* and *API Changes* are noted with a [directive and release note](https://matplotlib.org/devdocs/devel/api_changes.html#announce-changes-deprecations-and-new-features)
- [x] Documentation complies with [general](https://matplotlib.org/devdocs/devel/document.html#write-rest-pages) and [docstring](https://matplotlib.org/devdocs/devel/document.html#write-docstrings) guidelines

<!--We understand that PRs can sometimes be overwhelming, especially as the
reviews start coming in.  Please let us know if the reviews are unclear or
the recommended next step seems overly demanding, if you would like help in
addressing a reviewer's comments, or if you have been waiting too long to hear
back on your PR.-->

anntzer committed:
Parse {lua,xe}tex-generated dvi in dviread.

ksunden committed:
Zenodo v3.10.3

timhoffm committed:
Backport PR #30029: Update diagram in subplots_adjust documentation to clarify parameters

timhoffm committed:
Backport PR #30029: Update diagram in subplots_adjust documentation to clarify parameters

timhoffm committed:
Update doc/_embedded_plots/figure_subplots_adjust.py

fix flake8

timhoffm committed:
Update doc/_embedded_plots/figure_subplots_adjust.py

Fix flake8

rcomer committed:
Merge pull request #30034 from meeseeksmachine/auto-backport-of-pr-30029-on-v3.10.3-doc

Backport PR #30029 on branch v3.10.3-doc (Update diagram in subplots_adjust documentation to clarify parameters)

rcomer committed:
Merge pull request #30033 from meeseeksmachine/auto-backport-of-pr-30029-on-v3.10.x

Backport PR #30029 on branch v3.10.x (Update diagram in subplots_adjust documentation to clarify parameters)

useidemaisachola committed:
Backport PR #29895: The 'lines.markeredgecolor' now doesn't interfere on the color of errorbar caps

dstansby committed:
Merge pull request #30057 from QuLogic/auto-backport-of-pr-29895-on-v3.10.x

Backport PR #29895 on branch v3.10.x (The 'lines.markeredgecolor' now doesn't interfere on the color of errorbar caps)"

meeseeksmachine committed:
Backport PR #30079: FIX: cast legend handles to list (#30081)

Co-authored-by: hannah <story645@gmail.com>

timhoffm committed:
Backport PR #30096: Fix OffsetBox custom picker

timhoffm committed:
Merge pull request #30101 from meeseeksmachine/auto-backport-of-pr-30096-on-v3.10.x

rcomer committed:
Backport PR #30089: FIX: fix submerged margins algorithm being applied twice

QuLogic committed:
Merge pull request #30106 from meeseeksmachine/auto-backport-of-pr-30089-on-v3.10.x

Backport PR #30089 on branch v3.10.x (FIX: fix submerged margins algorithm being applied twice)

QuLogic committed:
Backport PR #30119: Add some types to _mathtext.py

QuLogic committed:
Backport PR #30119: Add some types to _mathtext.py

QuLogic committed:
Merge pull request #30122 from meeseeksmachine/auto-backport-of-pr-30119-on-v3.10.x

Backport PR #30119 on branch v3.10.x (Add some types to _mathtext.py)

meeseeksmachine committed:
Backport PR #30114: Fix _is_tensorflow_array. (#30120)

Co-authored-by: Antony Lee <anntzer.lee@gmail.com>

QuLogic committed:
Backport PR #30118: CI: Skip jobs on forks

QuLogic committed:
Backport PR #30118: CI: Skip jobs on forks

QuLogic committed:
Merge pull request #30142 from meeseeksmachine/auto-backport-of-pr-30119-on-v3.10.3-doc

Backport PR #30119 on branch v3.10.3-doc (Add some types to _mathtext.py)

QuLogic committed:
Merge pull request #30141 from meeseeksmachine/auto-backport-of-pr-30118-on-v3.10.3-doc

Backport PR #30118 on branch v3.10.3-doc (CI: Skip jobs on forks)

QuLogic committed:
Merge pull request #30140 from meeseeksmachine/auto-backport-of-pr-30118-on-v3.10.x

Backport PR #30118 on branch v3.10.x (CI: Skip jobs on forks)

timhoffm committed:
Backport PR #30144: js: Fix externally-controlled format strings

QuLogic committed:
Merge pull request #30146 from meeseeksmachine/auto-backport-of-pr-30144-on-v3.10.x

Backport PR #30144 on branch v3.10.x (js: Fix externally-controlled format strings)

AdwaithBatchu committed:
Implement arrow3d method in Axes3D

timhoffm committed:
Backport PR #30162: TST: Fix runtime error checking NaN input to format_cursor_data

QuLogic committed:
Merge pull request #30167 from meeseeksmachine/auto-backport-of-pr-30162-on-v3.10.x

Backport PR #30162 on branch v3.10.x (TST: Fix runtime error checking NaN input to format_cursor_data)

jklymak committed:
Backport PR #30180: DOC: expand polar example

jklymak committed:
Backport PR #30180: DOC: expand polar example

rcomer committed:
Backport PR #30180: DOC: expand polar example (#30190)

Co-authored-by: Jody Klymak <jklymak@gmail.com>

rcomer committed:
Merge pull request #30189 from meeseeksmachine/auto-backport-of-pr-30180-on-v3.10.x

Backport PR #30180 on branch v3.10.x (DOC: expand polar example)

timhoffm committed:
Backport PR #30212: [Doc]: fix bug in release notes for matplotlib v3.5.0 and v3.7.0

timhoffm committed:
Backport PR #30212: [Doc]: fix bug in release notes for matplotlib v3.5.0 and v3.7.0

timhoffm committed:
Merge pull request #30214 from meeseeksmachine/auto-backport-of-pr-30212-on-v3.10.3-doc

Backport PR #30212 on branch v3.10.3-doc ([Doc]: fix bug in release notes for matplotlib v3.5.0 and v3.7.0)

timhoffm committed:
Merge pull request #30213 from meeseeksmachine/auto-backport-of-pr-30212-on-v3.10.x

Backport PR #30212 on branch v3.10.x ([Doc]: fix bug in release notes for matplotlib v3.5.0 and v3.7.0)

tacaswell committed:
Backport PR #30198: Implement Path.__deepcopy__ avoiding infinite recursion

QuLogic committed:
Merge pull request #30217 from meeseeksmachine/auto-backport-of-pr-30198-on-v3.10.x

Backport PR #30198 on branch v3.10.x (Implement Path.__deepcopy__ avoiding infinite recursion)

meeseeksmachine committed:
Backport PR #30223: Polar log scale: fix inner patch boundary and spine location (#30227)

Co-authored-by: Tim Hoffmann <2836374+timhoffm@users.noreply.github.com>

meeseeksmachine committed:
Backport PR #30098: Fix label_outer in the presence of colorbars. (#30228)

Co-authored-by: Tim Hoffmann <2836374+timhoffm@users.noreply.github.com>

timhoffm committed:
Backport PR #30221: BUG: fix future incompatibility with Pillow 13

timhoffm committed:
Merge pull request #30229 from meeseeksmachine/auto-backport-of-pr-30221-on-v3.10.x

Backport PR #30221 on branch v3.10.x (BUG: fix future incompatibility with Pillow 13)

ianthomas23 committed:
Backport PR #30195: ci: Enable wheel builds on Python 3.14'

ci: Enable wheel builds on Python 3.14
(cherry picked from commit c805967aac1f63f8ee3fd12dd5f1ae9eb7ae56ee)

anntzer committed:
Backport PR #30209: Clean up Qt socket notifier to avoid spurious interrupt handler calls

QuLogic committed:
Merge pull request #30234 from meeseeksmachine/auto-backport-of-pr-30209-on-v3.10.x

Backport PR #30209 on branch v3.10.x (Clean up Qt socket notifier to avoid spurious interrupt handler calls)

timhoffm committed:
Backport PR #30236: Copy-edit the docstring of AuxTransformBox.

timhoffm committed:
Merge pull request #30240 from meeseeksmachine/auto-backport-of-pr-30236-on-v3.10.x

Backport PR #30236 on branch v3.10.x (Copy-edit the docstring of AuxTransformBox.)

timhoffm committed:
Backport PR #30246: chore: remove redundant words in comment

rcomer committed:
Merge pull request #30247 from meeseeksmachine/auto-backport-of-pr-30246-on-v3.10.x

Backport PR #30246 on branch v3.10.x (chore: remove redundant words in comment)

jklymak committed:
Backport PR #30244: DOC: Recommend to use bare Figure instances for saving to file

jklymak committed:
Backport PR #30244: DOC: Recommend to use bare Figure instances for saving to file

QuLogic committed:
Merge pull request #30251 from meeseeksmachine/auto-backport-of-pr-30244-on-v3.10.3-doc

Backport PR #30244 on branch v3.10.3-doc (DOC: Recommend to use bare Figure instances for saving to file)

QuLogic committed:
Merge pull request #30250 from meeseeksmachine/auto-backport-of-pr-30244-on-v3.10.x

Backport PR #30244 on branch v3.10.x (DOC: Recommend to use bare Figure instances for saving to file)

nrnavaneet committed:
Backport PR #30243: Fix FancyArrow rendering for zero-length arrows

jkseppan committed:
Merge pull request #30253 from QuLogic/auto-backport-of-pr-30243-on-v3.10.x

Backport PR #30243 on branch v3.10.x (Fix FancyArrow rendering for zero-length arrows)

jkseppan committed:
Backport PR #30237: Add explicit `**options: Any` for `add_subplot` method

anntzer committed:
Backport PR #30256: Time out in _get_executable_info

QuLogic committed:
Merge pull request #30259 from meeseeksmachine/auto-backport-of-pr-30256-on-v3.10.x

Backport PR #30256 on branch v3.10.x (Time out in _get_executable_info)

dstansby committed:
Backport PR #30233: Check that stem input is 1D

dstansby committed:
Merge pull request #30268 from meeseeksmachine/auto-backport-of-pr-30233-on-v3.10.x

Backport PR #30233 on branch v3.10.x (Check that stem input is 1D)

timhoffm committed:
Backport PR #30186: Fix figure legend when drawing stackplots

QuLogic committed:
Merge pull request #30269 from meeseeksmachine/auto-backport-of-pr-30186-on-v3.10.x

Backport PR #30186 on branch v3.10.x (Fix figure legend when drawing stackplots)

jkseppan committed:
Backport PR #30271: Reduce pause time in interactive timer test

QuLogic committed:
Backport PR #30286: Fix whitespace in _axes.py error message

QuLogic committed:
Backport PR #30283: changed the FAQ link to point to the correct path

gameroman committed:
Axes3D

gameroman committed:
axes3d

story645 committed:
Backport PR #30289: DOC: Fix build with pybind11 3

story645 committed:
Backport PR #30289: DOC: Fix build with pybind11 3

rcomer committed:
Merge pull request #30294 from meeseeksmachine/auto-backport-of-pr-30289-on-v3.10.3-doc

Backport PR #30289 on branch v3.10.3-doc (DOC: Fix build with pybind11 3)

QuLogic committed:
Merge pull request #30293 from meeseeksmachine/auto-backport-of-pr-30289-on-v3.10.x

Backport PR #30289 on branch v3.10.x (DOC: Fix build with pybind11 3)

QuLogic committed:
Merge pull request #30288 from meeseeksmachine/auto-backport-of-pr-30283-on-v3.10.x

Backport PR #30283 on branch v3.10.x (changed the FAQ link to point to the correct path)

QuLogic committed:
Merge pull request #30287 from meeseeksmachine/auto-backport-of-pr-30286-on-v3.10.x

Backport PR #30286 on branch v3.10.x (Fix whitespace in _axes.py error message)

dstansby committed:
Move release related docs to new sub-folder

gameroman committed:
#29341

dstansby committed:
Fix labeller

dstansby committed:
Move generate_credits.py

timhoffm committed:
Backport PR #30321: Fix type annotation for Axes.get_legend() to include None

dstansby committed:
Merge pull request #30326 from meeseeksmachine/auto-backport-of-pr-30321-on-v3.10.x

Backport PR #30321 on branch v3.10.x (Fix type annotation for Axes.get_legend() to include None)

lukashergt committed:
add tests for various corner cases of legends with `labelcolor` set to `linecolor`, e.g. ensuring that histograms with `step` histtype do not end up with an empty label

lukashergt committed:
fix legends with `labelcolor` set to `linecolor`: change ordering from `c, fc` to `mfc, fc, mec, ec, c`; check if `color` is empty, `none` or transparent and loop on if so

lukashergt committed:
account for the `classic` style in the mpl tests; need to specify `none` for both `c` and `ec` to ensure invisibility, in classic black edges are default

lukashergt committed:
remove unnecessary `loc=1` in new tests for `labelcolor_linecolor`

lukashergt committed:
use `mpl.colors.same_color` also for the `none` checks

lukashergt committed:
Apply suggestions from code review

Using helper for `'none'` string equality and using the fact that `to_rgba` always returns a 4-element tuple to simplify the code.

Co-authored-by: Tim Hoffmann <2836374+timhoffm@users.noreply.github.com>

lukashergt committed:
revert to original behaviour of having black legend labels for invisible handles; can pass empty string `" "` to achieve empty/invisible label instead; having a black legend label by default allows the creation of neutrally coloured dummy entries

lukashergt committed:
compactify tests for `legend_labelcolor_linecolor` as suggested in PR

tacaswell committed:
Backport PR #30344: Support fractional HiDPI in GTK4 backend

lukashergt committed:
remove unneeded calls to `plt.close`

lukashergt committed:
use tuple unpacking in `legend_labelcolor_linecolor` tests as suggested in PR review

ksunden committed:
Merge pull request #30292 from GameRoMan/auto-backport-of-pr-30237-on-v3.10.x

Backport PR #30237: Add explicit `**options: Any` for `add_subplot` m…

tacaswell committed:
Backport PR #28554: BLD: Enable wheels on Windows-on-ARM

QuLogic committed:
Backport PR #30345: qt: Use better devicePixelRatio event to refresh scaling

meeseeksmachine committed:
Backport PR #30327: FIX Update Axes limits from Axes.add_collection(... autolim=True) (#30351)

Co-authored-by: Thomas A Caswell <tcaswell@gmail.com>

ksunden committed:
Merge pull request #30277 from meeseeksmachine/auto-backport-of-pr-30271-on-v3.10.x

Backport PR #30271 on branch v3.10.x (Reduce pause time in interactive timer test)

ksunden committed:
Merge pull request #30350 from meeseeksmachine/auto-backport-of-pr-30344-on-v3.10.x

Backport PR #30344 on branch v3.10.x (Support fractional HiDPI in GTK4 backend)

ksunden committed:
Merge pull request #30353 from meeseeksmachine/auto-backport-of-pr-30345-on-v3.10.x

Backport PR #30345 on branch v3.10.x (qt: Use better devicePixelRatio event to refresh scaling)

ksunden committed:
Do not build 3.10 win-arm

Dependencies do not provide wheels, so do not try to build them ourselves (was failing because it was trying to install from nightlies, and no compatible version was provided).

tacaswell committed:
Merge pull request #30352 from meeseeksmachine/auto-backport-of-pr-28554-on-v3.10.x

Backport PR #28554 on branch v3.10.x (BLD: Enable wheels on Windows-on-ARM)

tacaswell committed:
Merge pull request #30356 from ksunden/auto-backport-of-pr-30195-on-v3.10.x

Manual Backport PR #30195 on branch v3.10.x (ci: Enable wheel builds on Python 3.14)

ksunden committed:
Fix PyPy sections (build for both 3.10 and 3.11

ksunden committed:
Update cibw version to 3.0.1

Matches the version on main

tacaswell committed:
Merge pull request #30357 from ksunden/3_10_4_cibw_pypy

CIBW updates: fix pypy sections, update cibw version

ksunden committed:
Github stats v3.10.4

ksunden committed:
Release prep v3.10.4

ksunden committed:
REL: v3.10.4

This is the fourth bugfix release of the 3.10.x series.

Included in this release is distributed wheels for Python 3.14
(including freethreaded) and Windows ARM.

There are also several smaller bugfixes.

ksunden committed:
REL: bump from v3.10.4 tag

ksunden committed:
switch version number to 3.10.5...

ksunden committed:
REL: v3.10.5

This is the fourth bugfix release of the 3.10.x series.

Included in this release is distributed wheels for Python 3.14
(including freethreaded) and Windows ARM.

There are also several smaller bugfixes.

ksunden committed:
REL: bump from 3.10.5 tag

ksunden committed:
Merge branch 'v3.10.3-doc' into v3.10.x

ksunden committed:
Zenodo v3.10.5

timhoffm committed:
ENH: Gracefully handle python-build-standalone ImportError with Tk.

Closes #30390.

QuLogic committed:
Qt: Fix HiDPI handling on X11/Windows

With X11, there is only ever a single scale, regardless of monitors as
in Wayland, so it's always the highest scale (i.e., it's 2 if you have
1.5&1-scaled monitors). Thus we get no change events and don't update
the internal scale. On Wayland though, as noted in the other issue from
Qt devs, you only get the fractional scale after the first expose, so
there's always at least one event there.

In the older/pre-#30345 code path, in `showEvent`, we'd call
`_update_screen` to set callbacks for its changes, and that _also_
called `_update_pixel_ratio`. In the new code, we don't have that
initial call, and because Wayland always has at least one event at
startup, it all seemed to work.

So just add the `_update_pixel_ratio` call in the new code path as well.
On X11, this will be the highest integer scale needed and never changes.
On Wayland, this will also be the rounded-up integer scale, but if using
fractional scaling, will change with a subsequent event to the correct
one. This does cause a few extra changes on startup, but should be more
consistent across platforms.

QuLogic committed:
TST: Simplify Qt pixel ratio test and ensure it tests initial scale

ksunden committed:
Update release guide

Add note about closing old milestones

Separate push steps for branch and tag

timhoffm committed:
DOC: Scale axis parameter

timhoffm committed:
Update lib/matplotlib/scale.py

Co-authored-by: Elliott Sales de Andrade <quantum.analyst@gmail.com>

Aaratrika-Shelly committed:
MNT/DOC: Deprecate anchor in Axes3D.set_aspect

r3kste commented on lib/mpl_toolkits/mplot3d/tests/test_axes3d.py:
```diff
@@ -2689,3 +2689,26 @@ def test_ndarray_color_kwargs_value_error():
+    start = [1, 2, 3]
+    end = [4, 5, 6]
+
+    ax.arrow3d(end, start,
+               arrowstyle="->, head_length=0.6, head_width=0.3", color='red')
```
I would prefer one or two more arrows plotted within the same figure, each with _different_ start, end  positions and also _different_ properties.

```python
ax.arrow3d(end1, start1, ...props1)
ax.arrow3d(end2, start2, ...props2)
```

r3kste commented on lib/mpl_toolkits/mplot3d/art3d.py:
```diff
@@ -1666,5 +1661,26 @@ def norm(x):
         colors[:, 3] = alpha
     else:
         colors = np.asanyarray(color).copy()
-
```
Revert these unnecessary changes (removal of empty lines) in `_shade_colors()`. It is better to not introduce unnecessary edits.

r3kste reviewed (changes requested):
@AdwaithBatchu This is looking pretty good already. Just some minor changes and we should be good to move on to examples/ documentation.

timhoffm committed:
Update lib/matplotlib/scale.py

Co-authored-by: hannah <story645@gmail.com>

doronbehar committed:
{Check,Radio}Buttons: Improve docs of label_props

timhoffm committed:
Apply suggestions from code review

Co-authored-by: hannah <story645@gmail.com>

timhoffm committed:
Update lib/matplotlib/scale.py

timhoffm committed:
Update lib/matplotlib/scale.py

timhoffm committed:
Merge pull request #30404 from timhoffm/doc-scale

DOC: Scale axis parameter

rcomer committed:
DOC: update Cartopy url

QuLogic committed:
Merge pull request #30414 from rcomer/cartopy-doc

DOC: update Cartopy url

QuLogic committed:
Backport PR #30414: DOC: update Cartopy url

QuLogic committed:
Merge pull request #30415 from meeseeksmachine/auto-backport-of-pr-30414-on-v3.10.x

Backport PR #30414 on branch v3.10.x (DOC: update Cartopy url)

dependabot[bot] committed:
Bump the actions group across 1 directory with 7 updates

Bumps the actions group with 7 updates in the / directory:

| Package | From | To |
| --- | --- | --- |
| [actions/checkout](https://github.com/actions/checkout) | `4.2.2` | `5.0.0` |
| [actions/download-artifact](https://github.com/actions/download-artifact) | `4.3.0` | `5.0.0` |
| [pypa/cibuildwheel](https://github.com/pypa/cibuildwheel) | `3.0.1` | `3.1.3` |
| [scientific-python/circleci-artifacts-redirector-action](https://github.com/scientific-python/circleci-artifacts-redirector-action) | `1.1.0` | `1.2.0` |
| [github/codeql-action](https://github.com/github/codeql-action) | `3.29.3` | `3.29.8` |
| [actions/cache](https://github.com/actions/cache) | `4.2.3` | `4.2.4` |
| [actions/first-interaction](https://github.com/actions/first-interaction) | `2.0.0` | `3.0.0` |



Updates `actions/checkout` from 4.2.2 to 5.0.0
- [Release notes](https://github.com/actions/checkout/releases)
- [Changelog](https://github.com/actions/checkout/blob/main/CHANGELOG.md)
- [Commits](https://github.com/actions/checkout/compare/11bd71901bbe5b1630ceea73d27597364c9af683...08c6903cd8c0fde910a37f88322edcfb5dd907a8)

Updates `actions/download-artifact` from 4.3.0 to 5.0.0
- [Release notes](https://github.com/actions/download-artifact/releases)
- [Commits](https://github.com/actions/download-artifact/compare/d3f86a106a0bac45b974a628896c90dbdf5c8093...634f93cb2916e3fdff6788551b99b062d0335ce0)

Updates `pypa/cibuildwheel` from 3.0.1 to 3.1.3
- [Release notes](https://github.com/pypa/cibuildwheel/releases)
- [Changelog](https://github.com/pypa/cibuildwheel/blob/main/docs/changelog.md)
- [Commits](https://github.com/pypa/cibuildwheel/compare/95d2f3a92fbf80abe066b09418bbf128a8923df2...352e01339f0a173aa2a3eb57f01492e341e83865)

Updates `scientific-python/circleci-artifacts-redirector-action` from 1.1.0 to 1.2.0
- [Release notes](https://github.com/scientific-python/circleci-artifacts-redirector-action/releases)
- [Commits](https://github.com/scientific-python/circleci-artifacts-redirector-action/compare/7eafdb60666f57706a5525a2f5eb76224dc8779b...839631420e45a08af893032e5a5e8843bf47e8ff)

Updates `github/codeql-action` from 3.29.3 to 3.29.8
- [Release notes](https://github.com/github/codeql-action/releases)
- [Changelog](https://github.com/github/codeql-action/blob/main/CHANGELOG.md)
- [Commits](https://github.com/github/codeql-action/compare/d6bbdef45e766d081b84a2def353b0055f728d3e...76621b61decf072c1cee8dd1ce2d2a82d33c17ed)

Updates `actions/cache` from 4.2.3 to 4.2.4
- [Release notes](https://github.com/actions/cache/releases)
- [Changelog](https://github.com/actions/cache/blob/main/RELEASES.md)
- [Commits](https://github.com/actions/cache/compare/5a3ec84eff668545956fd18022155c47e93e2684...0400d5f644dc74513175e3cd8d07132dd4860809)

Updates `actions/first-interaction` from 2.0.0 to 3.0.0
- [Release notes](https://github.com/actions/first-interaction/releases)
- [Commits](https://github.com/actions/first-interaction/compare/2d4393e6bc0e2efb2e48fba7e06819c3bf61ffc9...753c925c8d1ac6fede23781875376600628d9b5d)

---
updated-dependencies:
- dependency-name: actions/checkout
  dependency-version: 5.0.0
  dependency-type: direct:production
  update-type: version-update:semver-major
  dependency-group: actions
- dependency-name: actions/download-artifact
  dependency-version: 5.0.0
  dependency-type: direct:production
  update-type: version-update:semver-major
  dependency-group: actions
- dependency-name: pypa/cibuildwheel
  dependency-version: 3.1.3
  dependency-type: direct:production
  update-type: version-update:semver-minor
  dependency-group: actions
- dependency-name: scientific-python/circleci-artifacts-redirector-action
  dependency-version: 1.2.0
  dependency-type: direct:production
  update-type: version-update:semver-minor
  dependency-group: actions
- dependency-name: github/codeql-action
  dependency-version: 3.29.8
  dependency-type: direct:production
  update-type: version-update:semver-patch
  dependency-group: actions
- dependency-name: actions/cache
  dependency-version: 4.2.4
  dependency-type: direct:production
  update-type: version-update:semver-patch
  dependency-group: actions
- dependency-name: actions/first-interaction
  dependency-version: 3.0.0
  dependency-type: direct:production
  update-type: version-update:semver-major
  dependency-group: actions
...

Signed-off-by: dependabot[bot] <support@github.com>

dstansby committed:
Merge pull request #30416 from matplotlib/dependabot/github_actions/actions-755056202d

Bump the actions group across 1 directory with 7 updates

QuLogic committed:
Merge pull request #30304 from dstansby/move-rel-docs

Move release related docs to new sub-folder

r3kste commented on lib/mpl_toolkits/mplot3d/art3d.py:
```diff
@@ -1668,3 +1668,25 @@ def norm(x):
+        """Projects the points according to the renderer matrix."""
+        xs3d, ys3d, zs3d = self._verts3d
+        xs, ys, zs = proj3d.proj_transform(xs3d, ys3d, zs3d, self.axes.M)
+        self.set_positions((xs[0],ys[0]),(xs[1],ys[1]))
```
Looks like I missed this before, but there has to be spaces following the commas here.

```suggestion
        self.set_positions((xs[0], ys[0]), (xs[1], ys[1]))
```

r3kste commented on galleries/examples/mplot3d/arrows3d_demo.py:
```diff
@@ -0,0 +1,41 @@
+# test_arrows.py
```
The next step is to add 
1. an example and 
2. a what's new note.

Go through other examples in `galleries/examples/mplot3d`
and other what's new notes in `doc/users/next_whats_new` to get an idea on what is required.

r3kste reviewed (changes requested):
I think the tests are better now. I have added a comment explaining what needs to be done next for documentation.

timhoffm committed:
DOC: remove some usages of None as explicit defaults

In docs, we rather specify "optional" instead of "default: None" when
None is only used as a sentinel for "not given". We also specify
prefer concrete types or rc references over None usage if possible.

QuLogic committed:
Merge pull request #30422 from timhoffm/doc-none-as-not-given

DOC: remove some usages of None as explicit defaults

lukashergt committed:
limit the `AttributeError` exception block to the single relevant line calling `getattr`

anntzer committed:
Remove outdated reference to matplotlibbaselinemarker in tex sources.

The use of matplotlibbaselinemarker was removed in 2e351d1.

lukashergt committed:
minor reformatting to reduce line wrapping

timhoffm committed:
MNT: Registered 3rd party scales do not need an axis parameter anymore (#29358)

* MNT: Registered 3rd party scales do not need an axis parameter anymore

First step of #29349.

* Update doc/api/next_api_changes/deprecations/29358-TH.rst

Co-authored-by: Elliott Sales de Andrade <quantum.analyst@gmail.com>

---------

Co-authored-by: Elliott Sales de Andrade <quantum.analyst@gmail.com>

timhoffm committed:
Merge pull request #30425 from anntzer/mbm

Remove outdated reference to matplotlibbaselinemarker in tex sources.

anntzer committed:
Backport PR #30426: Fix a race condition in TexManager.make_dvi & make_png.

Includes an extra fix from PR #30431: Don't drop dpi in png filename
generation.

anntzer committed:
Fix a race condition in TexManager.make_dvi & make_png.

Previously, a race condition could occur if, while a process had called
make_tex (generating the tex file in the global cache) and was going to
call the latex subprocess (to generate the dvi file), another process
also called make_tex for the same tex string and started rewriting the
tex source.  In that case, the latex subprocess could see a partially
written (invalid) tex source.

Fix that by generating the tex source in a process-private temporary
directory, where the latex process is already going to run anyways.
(This is cheap compared to the latex subprocess invocation.)

Apply a similar strategy for make_png as well.

ksunden committed:
Apply suggested rewordings from review

timhoffm committed:
Merge pull request #30402 from ksunden/release_guide_update

Update release guide

ksunden committed:
Backport PR #30399: Qt: Fix HiDPI handling on X11/Windows

ksunden committed:
Merge pull request #30399 from QuLogic/fix-qt-x-hidpi

Qt: Fix HiDPI handling on X11/Windows

timhoffm committed:
Merge pull request #30426 from anntzer/mdr

Fix a race condition in TexManager.make_dvi.

anntzer committed:
Use pathlib in texmanager.

Only a few changes are needed.  Note that make_png returns an absolute
path so path joining is not even needed here.

Also restore the recently incorrectly removed usage of dpi in generating
the png filename.

Also drop an unrelated but outdated comment re: background handling,
which is obsolete since 8940c66 (look for `if hack: ...`).

doronbehar committed:
widgets.pyi: Fix type hints for label_props arguments

anntzer committed:
Use standard property alias machinery in contour().

Note that 'color(s)' remains special-cased because the ContourSet
constructor needs to do a fair bit of special wrangling for it.

tacaswell committed:
Merge pull request #30428 from meeseeksmachine/auto-backport-of-pr-30399-on-v3.10.x

Backport PR #30399 on branch v3.10.x (Qt: Fix HiDPI handling on X11/Windows)

QuLogic committed:
Merge pull request #30431 from anntzer/tp

Use pathlib in texmanager.

DominikStiller committed:
Document relative font sizes

DominikStiller committed:
Remove mentions of relative font sizes

timhoffm committed:
Merge pull request #30440 from DominikStiller/relative-font-sizes

Document relative font sizes

QuLogic committed:
Merge pull request #30434 from anntzer/pr30426-backport-with-fix

Backport PR #30426: Fix a race condition in TexManager.make_dvi & make_png.

QuLogic committed:
Merge pull request #30328 from lukashergt/fix-legend-labelcolor-linecolor

Fix legend `labelcolor=‘linecolor’` to handle various corner cases, e.g. step histograms and transparent markers

dependabot[bot] committed:
Bump github/codeql-action from 3.29.8 to 3.29.10 in the actions group

Bumps the actions group with 1 update: [github/codeql-action](https://github.com/github/codeql-action).


Updates `github/codeql-action` from 3.29.8 to 3.29.10
- [Release notes](https://github.com/github/codeql-action/releases)
- [Changelog](https://github.com/github/codeql-action/blob/main/CHANGELOG.md)
- [Commits](https://github.com/github/codeql-action/compare/76621b61decf072c1cee8dd1ce2d2a82d33c17ed...96f518a34f7a870018057716cc4d7a5c014bd61c)

---
updated-dependencies:
- dependency-name: github/codeql-action
  dependency-version: 3.29.10
  dependency-type: direct:production
  update-type: version-update:semver-patch
  dependency-group: actions
...

Signed-off-by: dependabot[bot] <support@github.com>

Commit:
corr typo axis -> axes

bhpcv252 committed:
added hardcoded colormap attributes for type checker support (issue: #30438)

tacaswell committed:
Backport PR #30444: Small correction of a typo in the galleries: axis instead of axes

tacaswell committed:
Merge pull request #30444 from aburlot/alan/fix-typo-axis-axes

Small correction of a typo in the galleries: axis instead of axes

timhoffm committed:
Merge pull request #30445 from meeseeksmachine/auto-backport-of-pr-30444-on-v3.10.x

Backport PR #30444 on branch v3.10.x (Small correction of a typo in the galleries: axis instead of axes)

dstansby committed:
Merge pull request #30412 from doronbehar/_Buttons--label_props.doc

{Check,Radio}Buttons: Improve docs of label_props

dstansby committed:
Backport PR #30412: {Check,Radio}Buttons: Improve docs of label_props

timhoffm committed:
Merge pull request #30448 from meeseeksmachine/auto-backport-of-pr-30412-on-v3.10.x

Backport PR #30412 on branch v3.10.x ({Check,Radio}Buttons: Improve docs of label_props)

dstansby committed:
Merge pull request #30441 from matplotlib/dependabot/github_actions/actions-153a9d9705

Bump github/codeql-action from 3.29.8 to 3.29.10 in the actions group

QuLogic committed:
Merge pull request #30446 from bhpcv252/fix-30438

Added hardcoded colormap attributes for type checker support

trygvrad committed:
MultiNorm class (#29876)

* MultiNorm class

This commit merges a number of commits now contained in  https://github.com/trygvrad/matplotlib/tree/multivariate-plot-prapare-backup , keeping only the MultiNorm class

* Apply suggestions from code review

Co-authored-by: Tim Hoffmann <2836374+timhoffm@users.noreply.github.com>

* Apply suggestions from code review

Co-authored-by: Tim Hoffmann <2836374+timhoffm@users.noreply.github.com>

* Updated input types for MultiNorm.__call__()

* improved error messages in MultiNorm

* updated types and errors for MultiNorm

* Apply suggestions from code review

Co-authored-by: Tim Hoffmann <2836374+timhoffm@users.noreply.github.com>

* Updated return type of MultiNorm

* Apply suggestions from code review

Co-authored-by: Tim Hoffmann <2836374+timhoffm@users.noreply.github.com>

* updated docstrings in MultiNorm

* Apply suggestions from code review

Co-authored-by: Elliott Sales de Andrade <quantum.analyst@gmail.com>

* changed so that vmin/vmax must be iterable

* Update colors.pyi

* change to handling of bad norm names in MultiNorm

* Update colors.py

* Apply suggestions from code review

Co-authored-by: Elliott Sales de Andrade <quantum.analyst@gmail.com>

* change minimum length of multiNorm to 1.

* Apply suggestions from code review

Co-authored-by: Elliott Sales de Andrade <quantum.analyst@gmail.com>

* Apply suggestions from @QuLogic

---------

Co-authored-by: Tim Hoffmann <2836374+timhoffm@users.noreply.github.com>
Co-authored-by: Elliott Sales de Andrade <quantum.analyst@gmail.com>

r3kste commented on galleries/examples/mplot3d/arrow3d.py:
```diff
@@ -0,0 +1,38 @@
+ax.arrow3d(end, start)
+
+end1 = np.array([1,2,3])
+# changing the properties of the arrow and arrow is drawn from default start
+# (0,0,0) to (1,2,3)
```
Change this to something like
```python
# Passing additional keyword arguments to control properties of the arrow.
# If the `start` parameter is not passed, the arrow is drawn from (0, 0, 0).
```

r3kste commented on galleries/examples/mplot3d/arrow3d.py:
```diff
@@ -0,0 +1,38 @@
+3D arrow plot
+=============
+
+Demonstrates plotting arrow in a 3D space. Here we plot two arrows from same start
+point to different end points, and change the properties of the second arrow by passing
+additional parameters other than the end point and start point to
+`~matplotlib.patches.FancyArrowPatch`.
+
+See `FancyArrowPatch documentation <https://matplotlib.org/stable/api/_as_gen/matplotlib.patches.FancyArrowPatch.html>`
+for how the kwargs are being processed in FancyArrowPatch.
```
1. No need to separately link the documentation of FancyArrowPatch. Just adding the reference to `pathches.FancyArrowPatch` is enough.
2. The second sentence is quite long. I have broken the second sentence into two to make it a bit simpler.

So something like this:

```python
"""
Demonstrates plotting arrows in a 3D space. 

Here we plot two arrows from the same start point to different
end points. The properties of the second arrow is changed by passing
additional parameters other than 'end' and 'start' to
`.patches.FancyArrowPatch`.
"""
```

r3kste commented on doc/users/next_whats_new/arrow3d.rst:
```diff
@@ -0,0 +1,4 @@
+New `arrow3d` Method
+--------------------
+
+This new feature allows users to plot 3D arrows in Matplotlib, enhancing the visualization capabilities for 3D data. The `arrow3d` method can be used to create an arrow in a 3D space, which is particularly useful for representing vectors or directional data.
```
I think this should **not** be a single continuous line. Add line breaks so that it doesn't cross the 88th column.

story645 committed:
factored out quick install index tab to seperate
file and added tab set to getting started guide

ianhi committed:
Add scroll capture functionality to WebAgg backend (#30403)

* Add scroll capture functionality to WebAgg backend

* lint

ksunden committed:
Merge up v3.10.5

QuLogic committed:
DOC: Move capture_scroll What's new note to new directory

tacaswell committed:
Merge pull request #30452 from QuLogic/fix-capture_scroll-note

DOC: Move capture_scroll What's new note to new directory

tacaswell committed:
DOC: add canonical uv update command in error message

tacaswell committed:
Merge pull request #30401 from ksunden/merge_up_v3.10.5

merge up v3.10.5

itsmeamitesh01 committed:
DOC: Fix typo in xlabel docstring

Logan-Pageler committed:
Added handling for undetermined home directory (#30454)

* Added handleing for undetermined home directory

* Add comment on the cause of the RuntimeError

---------

Co-authored-by: Tim Hoffmann <2836374+timhoffm@users.noreply.github.com>

tacaswell committed:
DOC: tweak suggestion for upgrading uv managed Python

Co-authored-by: Zanie Blue <contact@zanie.dev>

star1327p committed:
DOC: Correct a typo: confuzzlment -> confuzzlement

QuLogic committed:
Merge pull request #30455 from itsmeamitesh01/docs-fix-xlabel-typo

DOC: Fix typo in xlabel docstring

dstansby committed:
Merge pull request #30456 from star1327p/typo-in-guidelines

DOC: Correct a typo: confuzzlment -> confuzzlement

rcomer committed:
DOC: simplify hat graph example

timhoffm committed:
Fix line length

timhoffm committed:
Mentioned fix in recent python-build-standalone

timhoffm committed:
whitespace fix

timhoffm committed:
Backport PR #30451: doc: factor out quick install tab for reuse

timhoffm committed:
Backport PR #30451: doc: factor out quick install tab for reuse

timhoffm committed:
factored out quick install index tab to seperate (#30451)

file and added tab set to getting started guide

timhoffm committed:
Merge pull request #30460 from meeseeksmachine/auto-backport-of-pr-30451-on-v3.10.5-doc

Backport PR #30451 on branch v3.10.5-doc (doc: factor out quick install tab for reuse)

timhoffm committed:
Merge pull request #30461 from meeseeksmachine/auto-backport-of-pr-30451-on-v3.10.x

Backport PR #30451 on branch v3.10.x (doc: factor out quick install tab for reuse)

timhoffm committed:
Merge pull request #30459 from rcomer/hat-graph

DOC: simplify hat graph example

timhoffm committed:
Merge pull request #30433 from anntzer/ck

Use standard property alias machinery in contour().

trygvrad committed:
removed test_image_cursor_formatting()

This test does not test im.format_cursor_data() because it test it with 1d arrays, whereas the artist.format_cursor_data() should forward to colrizingArtist._format_cursor_data_override() but this only happens with 0d data. The correct behavior with nans is tested in the existing test test_format_cursor_data(), so the test removed here, even if working as intended, is superfluous.

anntzer committed:
Let ticklabels respect set_in_layout(False).

This is useful for manual finetuning of layout of static plots.

anntzer committed:
Let triage_tests support test modules with only figure_equals tests.

If there's only check_figure_equals tests in a module (e.g. test_pickle,
test_axis) then there's no directory in the baseline_images test, but
it still makes sense to run triage_tests in case of test failure.
Previously, the missing directory caused an error; this patches fixes
that.  To test this, change e.g. one of the check_figures_equal
tests in test_pickle to make it fail, run that test, and run
tools/triage_tests.py.

anntzer committed:
Deprecate redundant axes parameter to RadialLocator.

We can just directly fetch the relevant information from
`locator.axis.axes`; the case where `locator.axis.axes` was different
from `axes` was not tested and doesn't make much sense anyways.

The change also goes in the direction of having Locators carry less
state (and hopefully, in the future, not being associated with an axis,
but rather having the axis passed as argument to `__call__`).

trygvrad committed:
no need to senitize extrama in colorizer

the @vmin.setter in Normalize (and its subclasses) call _sanitize_extrema(), so there is no reason to also do so in colorizer

timhoffm committed:
Merge pull request #30468 from anntzer/tt

Let triage_tests support test modules with only figure_equals tests.

QuLogic committed:
Merge pull request #30470 from trygvrad/clean_colorizer_vmin

No need to sanitize extrema in Colorizer.set_clim

dstansby committed:
Add datetime test for ax.violin (#30384)

* Add datetime test for ax.violin

* Update lib/matplotlib/tests/test_datetime.py

Co-authored-by: Elliott Sales de Andrade <quantum.analyst@gmail.com>

---------

Co-authored-by: Tim Hoffmann <2836374+timhoffm@users.noreply.github.com>
Co-authored-by: Elliott Sales de Andrade <quantum.analyst@gmail.com>

QuLogic committed:
Merge pull request #30469 from anntzer/rla

Deprecate redundant axes parameter to RadialLocator.

brchristian committed:
Fix spelling error in `contains_branch_seperately` method name

Fixes #30474.

This PR fixes a long-standing spelling error in the `contains_branch_seperately` method name, which should be `contains_branch_separately`.

In order to maintain backwards compatibility, aliases maintain the old spelling for legacy use.

Type hints and internal references are updated to the new spelling accordingly.

QuLogic committed:
ci: Remove cibuildwheel override for win_arm64

Our dependencies now have wheels for that platform.

QuLogic committed:
ci: Remove cibuildwheel override for Python 3.14 testing

These all have 3.14 wheels available now that the release candidate is
out.

star1327p committed:
DOC: Fix text formatting of imshow_extent example (#30471)

* DOC: Make text font normal rather than italics

* DOC: Update the title of imshow_extent.py

ianthomas23 committed:
Merge pull request #30476 from QuLogic/cibw-default-arm64

ci: Remove cibuildwheel override for win_arm64/Py3.14

ianthomas23 committed:
Backport PR #30476: ci: Remove cibuildwheel override for win_arm64/Py3.14

rcomer committed:
MNT: correct _replacer docstring [ci doc]

story645 committed:
Merge pull request #30478 from rcomer/replacer-docstring

MNT: correct _replacer docstring

QuLogic committed:
Merge pull request #30467 from anntzer/til

Let ticklabels respect set_in_layout(False).

anntzer committed:
Clarify inset_locator.inset_axes demo.

... by clearly stating that this is the docs about
`inset_locator.inset_axes`, which is different from `Axes.inset_axes`.
Otherwise the last sentence ("Users should consider using
`Axes.inset_axes` instead") gets quite confusing.

timhoffm committed:
Merge pull request #30479 from anntzer/ilia

Clarify inset_locator.inset_axes demo.

ksunden committed:
Merge pull request #30394 from timhoffm/python-build-standalone-importerror

ENH: Gracefully handle python-build-standalone ImportError with Tk

ksunden committed:
Backport PR #30394: ENH: Gracefully handle python-build-standalone ImportError with Tk

ksunden committed:
Merge pull request #30477 from meeseeksmachine/auto-backport-of-pr-30476-on-v3.10.x

Backport PR #30476 on branch v3.10.x (ci: Remove cibuildwheel override for win_arm64/Py3.14)

tacaswell committed:
FIX: be more cautious about checking widget size

This can fail if the c++ side of the object has already been cleaned up
but the Python side is still lingering (due to refs in callbacks).

closes #29618

ksunden committed:
Merge pull request #30481 from meeseeksmachine/auto-backport-of-pr-30394-on-v3.10.x

Backport PR #30394 on branch v3.10.x (ENH: Gracefully handle python-build-standalone ImportError with Tk)

geofft committed:
doc: Update warnings about python-build-standalone (fixes #30409)

tacaswell committed:
CI: try skipping just released version of coverage

tacaswell committed:
CI: try pinning back pytest-rerunfailures instead

ksunden committed:
Backport PR #30484: FIX: be more cautious about checking widget size

ksunden committed:
Merge pull request #30484 from tacaswell/fix/qt_cleanup

FIX: be more cautious about checking widget size

tacaswell committed:
Backport PR #30486: doc: Update warnings about python-build-standalone

tacaswell committed:
Backport PR #30486: doc: Update warnings about python-build-standalone

tacaswell committed:
Merge pull request #30486 from geofft/python-build-standalone-tkinter-doc

doc: Update warnings about python-build-standalone

ksunden committed:
Merge pull request #30487 from meeseeksmachine/auto-backport-of-pr-30484-on-v3.10.x

Backport PR #30484 on branch v3.10.x (FIX: be more cautious about checking widget size)

ksunden committed:
Merge pull request #30488 from meeseeksmachine/auto-backport-of-pr-30486-on-v3.10.x

Backport PR #30486 on branch v3.10.x (doc: Update warnings about python-build-standalone)

ksunden committed:
Merge pull request #30489 from meeseeksmachine/auto-backport-of-pr-30486-on-v3.10.5-doc

Backport PR #30486 on branch v3.10.5-doc (doc: Update warnings about python-build-standalone)

ksunden committed:
Github stats v3.10.6

ksunden committed:
Merge branch 'v3.10.5-doc' into v3.10.x

ksunden committed:
Release prep v3.10.6

ksunden committed:
REL: v3.10.6

This is a bugfix release in the 3.10.x series.

Highlights from this release include:

- Fix regression of hi-dpi support for Qt
- Fix race condition in TexManager.make_dvi & make_png
- Various documentation and other bugfixes

ksunden committed:
REL: bump from v3.10.6

ksunden committed:
Zenodo v3.10.6

ksunden committed:
Merge branch 'v3.10.x' into merge_up_v3.10.6

star1327p committed:
DOC: pytz link should be from PyPI

QuLogic committed:
Merge pull request #30492 from star1327p/dependencies

DOC: pytz link should be from PyPI

MengAiDev committed:
Fix Line3DCollection with autolim=True for lines of different lengths (#30423)

* Fix Line3DCollection with autolim=True for lines of different lengths

When using Line3DCollection with autolim=True, the code was trying to
convert ragged arrays (lines with different numbers of points) directly
to a numpy array, which would fail. This fix extracts coordinates
separately to avoid the issue.

Fixes #30418

* fix lint

* fix lint

* Add test for Line3DCollection with autolim=True and ragged arrays

* move test

* Update lib/mpl_toolkits/mplot3d/tests/test_axes3d.py

Co-authored-by: Elliott Sales de Andrade <quantum.analyst@gmail.com>

* Update lib/mpl_toolkits/mplot3d/tests/test_axes3d.py

Co-authored-by: Elliott Sales de Andrade <quantum.analyst@gmail.com>

---------

Co-authored-by: Elliott Sales de Andrade <quantum.analyst@gmail.com>

r3kste commented on doc/users/next_whats_new/arrow3d.rst:
```diff
@@ -0,0 +1,7 @@
+This new feature allows users to plot 3D arrows in Matplotlib,
+enhancing the visualization capabilities for 3D data.
+The `arrow3d` method can be used to create an arrow in a 3D space,
+which is particularly useful for representing vectors or directional data.
```
This feels too repetitive. Both sentences convey exactly the same meaning, so you could simplify to:

```suggestion
The new ``arrow3d`` method allows users to plot 3D arrows,
which can be used for representing vectors or directional data.
```

r3kste commented on doc/users/next_whats_new/arrow3d.rst:
```diff
@@ -0,0 +1,7 @@
+New `arrow3d` Method
```
In .rst file
1. To denote variables, use double ticks ( **``** )
2. Single ticks are used to add a reference to mentioned variable.

```suggestion
New ``arrow3d`` method for `~.Axes3D`
```

r3kste commented on lib/mpl_toolkits/mplot3d/axes3d.py:
```diff
@@ -4062,6 +4062,46 @@ def stem(self, x, y, z, *, linefmt='C0-', markerfmt='C0o', basefmt='C3-',
+        arrow = art3d.Arrow3D(xs, ys, zs, **arrow_props)
+        self.add_artist(arrow)
+        self.auto_scale_xyz(xs, ys, zs, had_data)
+        return arrow
```
Add line breaks between different code blocks to make it easier to distinguish between.

```suggestion
        had_data = self.has_data()

        if start is None:
            start = np.zeros_like(end)
        if np.shape(end) != (3,):
            raise ValueError("end must be an array of length 3")
        if np.shape(start) != (3,):
            raise ValueError("start must be an array of length 3")

        # Set default arrow properties and update with any additional keyword args
        arrow_props = dict(
            mutation_scale=20, arrowstyle="-|>", shrinkA=0, shrinkB=0
        )
        arrow_props.update(kwargs)

        xs, ys, zs = [start[0], end[0]], [start[1], end[1]], [start[2], end[2]]
        arrow = art3d.Arrow3D(xs, ys, zs, **arrow_props)
        self.add_artist(arrow)
        self.auto_scale_xyz(xs, ys, zs, had_data)

        return arrow
```

r3kste commented on lib/mpl_toolkits/mplot3d/tests/test_axes3d.py:
```diff
@@ -2689,3 +2689,30 @@ def test_ndarray_color_kwargs_value_error():
+    start2 = [2, 5, 7]
+    end2 = [4, 6, -8]
+    ax.arrow3d(end2, start2, color='violet', ls='--',
+               lw=2)
```
Why unnecessary line break?

```suggestion
    ax.arrow3d(end2, start2, color='violet', ls='--', lw=2)
```

r3kste reviewed (changes requested):
Looks very good now @AdwaithBatchu 

I have just a few minor comments with simple changes, which is basically nitpicking at this point. Once this is done, I will build docs to verify everything seems fine.

jonathimer committed:
Add Linux Foundation Health Score badge to README

Congrats! We have onboarded matplotlib to LFX Insights, the Linux Foundation's platform for monitoring the world's most critical open-source projects.

https://insights.linuxfoundation.org/project/matplotlib

AasmaGupta committed:
Fix SVG rendering error in def update_background (#30490)

Check if saving before updating blitting background

---------

Co-authored-by: AASMA GUPTA <aasmagupta@Aasmas-Mac.local>
Co-authored-by: Tim Hoffmann <2836374+timhoffm@users.noreply.github.com>

timhoffm committed:
Merge pull request #30505 from jonathimer/patch-1

Add Linux Foundation Health Score badge to README

ksunden committed:
Revert 3.9 link

timhoffm committed:
Merge pull request #30475 from brchristian/fix_method_name_typo

Fix spelling error in `contains_branch_separately` method name

QuLogic committed:
Merge pull request #30491 from ksunden/merge_up_v3.10.6

merge up v3.10.6

scottshambaugh committed:
Merge pull request #30408 from Aaratrika-Shelly/fix/axes3d-set-aspect

MNT/DOC: Deprecate anchor in Axes3D.set_aspect

rcomer committed:
Update syntax for PR welcome workflow

It looks like the syntax for this changed at https://github.com/actions/first-interaction/pull/311

story645 committed:
Merge pull request #30510 from rcomer/pr-greet

Update syntax for PR welcome workflow

timhoffm committed:
Merge pull request #29939 from anntzer/luadviparse

Parse {lua,xe}tex-generated dvi in dviread.

dstansby committed:
Merge pull request #30465 from trygvrad/remove-test_image_cursor_formatting

removed test_image_cursor_formatting()
